File: python/open_guardrail/guards/age_appropriate_content.py
"""Enforce content filtering for K-12 audiences."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_GRADE_CTX = re.compile(
    r"\b(?:(?:for\s+)?(?:1st|2nd|3rd"
    r"|[4-8]th)\s+graders?"
    r"|elementary|K-?12"
    r"|middle\s+school"
    r"|kindergarten)\b",
    re.IGNORECASE,
)
_INAPPROPRIATE = re.compile(
    r"\b(?:violen(?:ce|t)|murder"
    r"|kill(?:ing)?|drugs?|cocaine"
    r"|heroin|meth|explicit"
    r"|sex(?:ual)?|porn"
    r"|gambl(?:e|ing)|alcohol"
    r"|beer|wine|liquor"
    r"|cigarette|tobacco"
    r"|weapon|firearm|suicide)\b",
    re.IGNORECASE,
)


def _estimate_grade_level(text: str) -> int:
    sentences = [
        s
        for s in re.split(r"[.!?]+", text)
        if s.strip()
    ]
    if not sentences:
        return 0
    words = [
        w for w in text.split() if w
    ]
    if not words:
        return 0
    avg_wps = len(words) / len(sentences)
    total_syl = sum(
        max(
            1,
            len(
                re.findall(
                    r"[aeiouy]+", w, re.I
                )
            ),
        )
        for w in words
    )
    avg_syl = total_syl / len(words)
    return round(
        0.39 * avg_wps
        + 11.8 * avg_syl
        - 15.59
    )


class _AgeAppropriateContent:
    def __init__(
        self,
        *,
        action: str = "block",
        max_grade_level: int = 8,
    ) -> None:
        self.name = "age-appropriate-content"
        self.action = action
        self.max_grade = max_grade_level

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []
        grade = _estimate_grade_level(text)
        if grade > self.max_grade:
            issues.append(
                f"Reading level (grade {grade})"
                f" exceeds max"
                f" (grade {self.max_grade})"
            )
        has_ctx = bool(
            _GRADE_CTX.search(text)
        )
        has_bad = bool(
            _INAPPROPRIATE.search(text)
        )
        if has_bad and (
            has_ctx or self.max_grade <= 8
        ):
            issues.append(
                "Age-inappropriate content"
                " detected for target audience"
            )
        elapsed = (
            time.perf_counter() - start
        ) * 1000
        if not issues:
            return GuardResult(
                guard_name=(
                    "age-appropriate-content"
                ),
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )
        return GuardResult(
            guard_name="age-appropriate-content",
            passed=False,
            action=self.action,
            message="; ".join(issues),
            latency_ms=round(elapsed, 2),
            details={
                "grade_level": grade,
                "max_grade_level": self.max_grade,
                "issues": issues,
                "has_grade_context": has_ctx,
                "has_inappropriate": has_bad,
            },
        )


def age_appropriate_content(
    *,
    action: str = "block",
    max_grade_level: int = 8,
) -> _AgeAppropriateContent:
    return _AgeAppropriateContent(
        action=action,
        max_grade_level=max_grade_level,
    )

File: packages/guards/src/age-appropriate-content.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';
interface AgeAppropriateContentOptions { action: 'block' | 'warn'; maxGradeLevel?: number; }
const GRADE_CTX = /\b(?:(?:for\s+)?(?:1st|2nd|3rd|[4-8]th)\s+graders?|elementary|K-?12|middle\s+school|kindergarten)\b/i;
const INAPPROPRIATE = /\b(?:violen(?:ce|t)|murder|kill(?:ing)?|drugs?|cocaine|heroin|meth|explicit|sex(?:ual)?|porn|gambl(?:e|ing)|alcohol|beer|wine|liquor|cigarette|tobacco|weapon|firearm|suicide)\b/i;
function estimateGradeLevel(text: string): number {
  const sentences = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
  if (sentences.length === 0) return 0;
  const words = text.split(/\s+/).filter(w => w.length > 0);
  const avgWordsPerSentence = words.length / sentences.length;
  const syllableCount = words.reduce((sum, w) => sum + Math.max(1, (w.match(/[aeiouy]+/gi) || []).length), 0);
  const avgSyllables = syllableCount / words.length;
  return Math.round(0.39 * avgWordsPerSentence + 11.8 * avgSyllables - 15.59);
}
export function ageAppropriateContent(options: AgeAppropriateContentOptions): Guard {
  const maxGrade = options.maxGradeLevel ?? 8;
  return { name: 'age-appropriate-content', version: '0.1.0', description: 'Enforce content filtering for K-12 audiences', category: 'content',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const issues: string[] = [];
      const gradeLevel = estimateGradeLevel(text);
      if (gradeLevel > maxGrade) issues.push(`Reading level (grade ${gradeLevel}) exceeds max (grade ${maxGrade})`);
      const hasGradeCtx = GRADE_CTX.test(text);
      const hasInappropriate = INAPPROPRIATE.test(text);
      if (hasInappropriate && (hasGradeCtx || maxGrade <= 8)) issues.push('Age-inappropriate content detected for target audience');
      const triggered = issues.length > 0;
      return { guardName: 'age-appropriate-content', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? issues.join('; ') : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { gradeLevel, maxGradeLevel: maxGrade, issues, hasGradeContext: hasGradeCtx, hasInappropriateContent: hasInappropriate } : undefined,
      };
    },
  };
}
